src/rules.jl

```diff
@@ -96,3 +96,110 @@ true
 See also: [`frule`](@ref), [`@scalar_rule`](@ref)
 """
 rrule(::Any, ::Vararg{Any}; kwargs...) = nothing
+
+
+#######################################################################
+# Infastructure to support generating overloads from rules.
+
+const NEW_RRULE_HOOKS = Function[]
+const NEW_FRULE_HOOKS = Function[]
+_hook_list(::typeof(rrule)) = NEW_RRULE_HOOKS
+_hook_list(::typeof(frule)) = NEW_FRULE_HOOKS
+
+"""
+    on_new_rule(sig->eval...), frule | rrule)
+"""
+function on_new_rule(hook_fun, rule_kind)
+    # get all the existing rules
+    ret = map(_rule_list(rule_kind)) do method
+        sig = _primal_sig(rule_kind, method)
+        _safe_hook_fun(hook_fun, sig)
+    end
+
+    # register hook for new rules
+    push!(_hook_list(rule_kind), hook_fun)
+    return ret
+end
+
+function __init__()
+    push!(Base.package_callbacks, pkgid -> refresh_rules())
+    push!(Base.include_callbacks, (mod, filename) -> refresh_rules())
+end
+
+
+"""
+    _rule_list(frule | rrule)
+
+Returns a list of all the methods of the currently defined rules of the given kind.
+Excluding the fallback rule that returns `nothing` for every input.
+"""
+_rule_list(rule_kind) = (m for m in methods(rule_kind) if m.module != @__MODULE__)
+# ^ The fallback rules are the only rules defined in ChainRules core  so that is how we skip them.
+
+
+
+const LAST_REFRESH_RRULE = Ref(0)
+const LAST_REFRESH_FRULE = Ref(0)
+last_refresh(::typeof(frule)) = LAST_REFRESH_FRULE
+last_refresh(::typeof(rrule)) = LAST_REFRESH_RRULE
+
+"""
+    refresh_rules(); refresh_rules(frule); refresh_rules(rrule)
```

Comment:
we seem to have the `frule | rrule` style elsewhere, so may as well use it here too
```suggestion
    refresh_rules()
    refresh_rules(frule | rrule)
```

Reply:
also valid is:
```suggestion
    refresh_rules([frule | rrule])
```

per man-page convention